### PR TITLE
Make RevapiJavaTask inputs deterministic

### PR DESCRIPTION
**Pull request by CRogers**
## Before this PR
It is impossible to add caching to revapi as we have no way of knowing which version of the API we going to test against before we actually run the task (since the latest git tag may not have been published yet).

## After this PR
==COMMIT_MSG==
The old version is now an input property on the reavpi task, which means the actual task itself is deterministic.
We therefore have to compute which old version we're comparing against each time.
==COMMIT_MSG==

### CODE REVIEW

File: src/main/java/com/palantir/gradle/revapi/OldApiConfigurations.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.artifacts.result.DependencyResult;
+import org.gradle.api.artifacts.result.UnresolvedDependencyResult;
+
+final class OldApiConfigurations {
```
Comment (CRogers) [inline src/main/java/com/palantir/gradle/revapi/OldApiConfigurations.java:31]: All this code is moved from this file: https://github.com/palantir/gradle-revapi/pull/108/files#diff-84a5460c13e2bc9d571185e418f5bf68L197

File: src/main/java/com/palantir/gradle/revapi/ResolveOldApiVersion.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.stream.Collectors;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.provider.Provider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("Duplicates")
+final class ResolveOldApiVersion {
```
Comment (CRogers) [inline src/main/java/com/palantir/gradle/revapi/ResolveOldApiVersion.java]: This code is almost identical to that from: https://github.com/palantir/gradle-revapi/pull/108/files#diff-84a5460c13e2bc9d571185e418f5bf68L124, just made static

File: src/main/java/com/palantir/gradle/revapi/OldApiConfigurations.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                .filter(dependencyResult -> dependencyResult instanceof UnresolvedDependencyResult)
+                .map(dependencyResult -> (UnresolvedDependencyResult) dependencyResult)
+                .map(UnresolvedDependencyResult::getFailure)
+                .collect(Collectors.toList());
+
+        if (resolutionFailures.isEmpty()) {
+            return configuration.resolve();
+        }
+
+        throw new CouldNotResolveOldApiException(oldVersion, resolutionFailures);
```
Comment (dansanduleac) [inline src/main/java/com/palantir/gradle/revapi/OldApiConfigurations.java:72]: it's weird that we're passing in this `oldVersion` only to put it back into an exception, so it doesn't actually affect the behaviour of this method).

Perhaps we could construct the configuration in here, and have the function take a name/prefix for the configuration?

essentially putting this code https://github.com/palantir/gradle-revapi/pull/108/files#diff-1da680adc77435cdeb6de908e5415d62R84-R90 inside the function.

File: src/main/java/com/palantir/gradle/revapi/PreviousVersionResolutionHelpers.java [hunk 4]
```diff
@@ -58,6 +58,13 @@ public String toString() {
         }
     }
 
+    /**
+     * When the version of the local java project is higher than the old published dependency and has the same
+     * group and name, gradle silently replaces the published external dependency with the project dependency
+     * (see https://discuss.gradle.org/t/fetching-the-previous-version-of-a-projects-jar/8571). This happens on
+     * tag builds, and would cause the publish to fail. Instead we, change the group for just this thread
```
Comment (dansanduleac) [inline src/main/java/com/palantir/gradle/revapi/PreviousVersionResolutionHelpers.java]: Instead, we

File: src/main/java/com/palantir/gradle/revapi/RevapiJavaTask.java [hunk 5]
```diff
@@ -67,8 +71,18 @@
         return newApiJars;
     }
 
+    public final Property<GroupAndName> getOldGroupAndName() {
```
Comment (dansanduleac) [inline src/main/java/com/palantir/gradle/revapi/RevapiJavaTask.java]: you need to mark these with `@Input` (`getOldVersion` too)

Comment (dansanduleac) [reply src/main/java/com/palantir/gradle/revapi/RevapiJavaTask.java]: ah I see that one of the implementations has `getOldGroupNameVersion` which is marked as input.

Why don't we delete these two then and pull up `getOldGroupNameVersion` into this abstract class, and extract the group+name and version respectively from that, so there's a single source of truth?

Comment (CRogers) [reply src/main/java/com/palantir/gradle/revapi/RevapiJavaTask.java]: I'm wasn't planning to mark anything as `@Input` until it actually caches properly :p

File: src/main/java/com/palantir/gradle/revapi/OldApiConfigurations.java (hunk 3 above)
Comment (CRogers) [reply src/main/java/com/palantir/gradle/revapi/OldApiConfigurations.java:72]: I'm still passing in `oldVersion` here, but this method is now private, so I don't think it's as bad (and inlining this method would look ugly, I think).

**Review by dansanduleac — APPROVED** [review]

**Pull request merged**